### DISCUSSION TITLE
Broken parameter types for in-component navigation guards supplied to `defineComponent`

### DISCUSSION THREAD

**Discussion by ghost** [vuejs/router#2332] [Help and Questions]
### Reproduction

https://codesandbox.io/s/vue-router-4-reproduction-s1sqc

### Steps to reproduce the bug

View intellisense for parameters in `beforeRouteEnter`

### Expected behavior

Type parameters in `beforeRouteEnter`:
- `to: RouteLocationNormalizedGeneric`
- `from: RouteLocationNormalizedGeneric`
- `next: NavigationGuardNext`

### Actual behavior

Type parameters in `beforeRouteEnter`:
- `to: any`
- `from: any`
- `next: any`

**Comment by posva** [comment]
It [works](https://play.vuejs.org/#eNqlVVtv2jAU/iteNgnQIKa06kOWonZVtW7apeqmvSx7CIkBt4lt2Q6lQvnvO7ZzBdqXCaHE5/Kdy3d8svPymDL/QXmBR3PBpUY7lEgSa3IlBCrRUvIcDTYFGUSsMpC80ERWGh+7Y6s2frUuFsJ3vhFrUIfwH0UMIb9QZOjc3TnnBdPDwVtwG4y8sVe5Q26hJrnIwH9uDMP1yfyWZBk3wd6EGI5WLOwDXpSWnK3m14WUhFUZIxHrdRDiSod2O/TOKvxlkWV3oERlaWGwwwlZvKkB722WXyl7RJpfRB6OvPknDu/oluckxK3+RY94AZLG7cqcDvxCXMcMDS99rN+UPCFc2VXqEDeNgX65Xh5l8xvJuXy+pUrDY1wJHW6X5UnDZsOnKdCGrkmtBRWzNe2moJ5dI2lHgDNVsaHQBfpjStk5XtAADyAtDmAMKAvasOW4b2b72LdtYxvjv3uhJITq1jvcGcC1a0VwrD/DkY3pEoXXEsYzYmRrC03JMi6yGhu63u2HGVWVSCo0UkQXAmUxWwH7WgHzhi6nhFc47M30bF4jwUTPDsjttfP/4jRQRwM5Qid5LGCSOINQtmFRpYAIAbISI2tnxogjb621UAHGBROPKx84wq3F5Zl/6k9xCk3uSH2i8slC8idlRzfybO8N9iUY4ZRsNOeZmsSCvhTiwPDy3D/3T3BGFxjQMWUp2VpsAw2XvIQytYIZWdLVXpFmrGhG5A+hKcxQr9gYVs7TFyvTsiBNosmaJI9H5A9q61K+kwSq25BOcTqWKwL7wKhvfn4nW3hvlDlPiwysX1HeE8WzwuTozD4WLIW0O3Y228+WM8pWv9TNVhOm6qJMorYb1t4Sef1K6W26p/5Zp4tKP2dE+YkyOweW1xiZzeT8FlymRAZoJrYIkqUpejudTj8YVQ5wlE0WXGueB+hkKrZWLuI0hWQbCUSJGMCiOYrRe/hb4Mo7I0u4/F3L9awbuYVvE0iSpJNAgKbwm1UIUI5pwd796t+sZq3CGqDMtswuod6nEsAO1sWevVtCC7Lk0q2lG2ZWk+ajuuea+6JuNOyf1650SpsPlcq4bj4STt653RHzyn+9TLMy)

![Screenshot 2024-08-06 at 17 07 29](https://github.com/user-attachments/assets/67a3a4ae-cd65-4315-b2ed-30361267d83b)

PS: please don't open issues with incomplete reproductions, it's not showcasing what you say.

**Reply by ghost** [reply]
Thanks for the quick reply! I appreciate your patience on this, it is my very first time ever contributing to any open source project.

I did try this in that playground and could not recreate it, even with the versions matching the project:
- TS: 5.5.4
- Vue: 3.4.36
- Vue Router:
  - 4.4.0 (without issue)
  - 4.4.1 (introduces issue)
  - •••
  - 4.4.3 (still has issue)

My next thought was that it was an issue with vue-language-tools, but that did not give me any clues.

Finally, I decided to play around with the extensions and stumbled across something, so I've opened a [PR](https://github.com/vuejs/router/pull/2333) which includes a demonstration of the issue as well as the partial reversion we could do to address it.

Thanks for all your hard work, @posva!

**Reply by posva** [reply]
Thanks but the reproduction is crucial to move this forward

**Reply by ghost** [reply]
Makes sense, I'll try again!

I assume you were able to get a reproduction for [the change](https://github.com/vuejs/router/commit/a495ce0e1d545d521673d21af43e63a761b53e27) that would be partially reverted since _that_ change was accepted, right? I couldn't find that sandbox, all I see are screenshots. Could you point me towards that sandbox or whatever it was that let you reproduce?
